### PR TITLE
Rover: Direct-rotation sail control

### PR DESCRIPTION
**Pull request by arikrupnik**
Allows a sailboat configuration where a servo directly rotates the mast, rather than controlling its AoA indirectly with a sheet or an elevator-like surface.

### CODE REVIEW

File: Rover/AP_MotorsUGV.cpp
Comment (IamPete1) [inline Rover/AP_MotorsUGV.cpp:246]:
```suggestion
void AP_MotorsUGV::set_mast_rotation(float mast_rotation)
```

**Comment by IamPete1** [comment]
looks good so far, just need the the magic function to set it.

You will also need to add it to `get_pilot_desired_mainsail`

This is slightly more complex than the existing outputs, they both have a fixed position to de-power. You will have to actively drive this to point into the wind to de-power. You might have to add some code paths for this, `get_stop_sail` or something, can just return 0 for the two existing outputs.

File: Rover/AP_MotorsUGV.cpp
Comment (arikrupnik) [reply Rover/AP_MotorsUGV.cpp]: Thank you for noticing this. Fixed.

Is the process in Ardupilot to squash commits before merging into master? Asking for commit messages for little fixes like this.

Comment (IamPete1) [reply Rover/AP_MotorsUGV.cpp]: Yeah, you should have one (or more) commit per directory. So you would have `Rover: something` and `SRV_Channels: something else`. There is a wiki page on how to interactive rebase. https://ardupilot.org/dev/docs/git-interactive-rebase.html#interactive-rebase-cleaning-up-commits, no real requirement to keep it tidy while its a WIP tho, feel free to wait until its working nice and then tidy up at the end.

**Comment by arikrupnik** [comment]
Thank you for pointing out `get_pilot_desired_mainsail`. I'm not sure what it represents. Is it the servo position that the operator would request if the boat were in manual mode? In this case, I think it's exactly as `wingsail_out`, both in the absence of the input and in its presence. Forcing `mast_rotation` to 0.0f would orient the sail parallel to the hull. This is as close to completely depowering as there is. Reaching wind would heel the boat without making any headway, and headwind would push her backwards only as much as the sail's Cd. I think actively driving the sail to point into the wind can reduce heel (and leeway), but would allow headway to the extent that the alignment with wind is imperfect. Perfect alignment with keel guarantees zero headway.

**Comment by IamPete1** [comment]
> Thank you for pointing out `get_pilot_desired_mainsail`. I'm not sure what it represents. Is it the servo position that the operator would request if the boat were in manual mode?

Excatly.

I'm not sure about just fixing the wing to parallel to the hull, I think your just going to get knocked about with excessive heel. It also means you have to go via a high angle of attack to get to a lower one, so setting off and stopping might get quite exciting.  I also don't think you would actually stop, I suspect you might just end still going forward a bit and slipping sideways alot.

Feel free to try it IRL and prove me wrong, we can always add pointing the wing into the wind as a future improvement.

**Comment by arikrupnik** [comment]
I can try to take some videos that show this. I'm still waiting for my wind vane, but this is something I can demonstrate under manual control. I have the sail control on a knob on my TX, so that rotating the knob clockwise rotates the mast clockwise. Middle detent on the knob is the "off" position for this sail.

Mind you, with a sail winch, sheeting all the way out depowers you only in as much as you're going upwind. If the wind happens to be from behind, you'll get all the power :=) The future improvement you speak of could actively manage the sail angle for mast rotation and for the winch. Sort of the opposite of what the comments call "ideal angle" for the winch.

Outside of that, is there anything else I'm forgetting before I get to the magic? (it's not actually that magical, but I do need the hardware).

**Comment by IamPete1** [comment]
> Outside of that, is there anything else I'm forgetting before I get to the magic? (it's not actually that magical, but I do need the hardware).

Its looking good. Just need the extra few lines ;). I suspect you will have to add a offset param. We could assume that the servo range is mapped -180 to +180 relative to the hull, then just add a offset param to remove the need to do that mechanically. Should be able to re-use the ideal sail angle param to set the sail relative to the wing to go forwards (and backwards).

File: Rover/AP_MotorsUGV.h
Comment (rmackay9) [inline Rover/AP_MotorsUGV.h:88]: this argument looks like it might be a cut-and-paste error

**Comment by arikrupnik** [comment]
Thank you for your encouragement, Peter.

I'm not sure I understand fully your meaning about the offset parameter. Are you talking about constraining the range of rotation or setting the middle point? I would have thought you can control both with `SERVOx_MIN/MAX/TRIM` parameters.

I'm planning to use `SAIL_ANGLE_IDEAL` like you suggest, and `SAIL_ANGLE_MAX` to map `SERVOx_MIN/MAX` to degrees of rotation. On my boat the geometry works out so the sail can rotate only about +-85 degrees. You never need to rotate more than +-90 degrees, even if you want to go backwards. I really wasn't panning on having the boat reverse, but if you really want to, you simply put the sail sharp edge into the wind--it works better than you might expect; the only major downside is the sharper stall. I'm planning to ignore `SAIL_ANGLE_MIN`.

The ideal angle applies when sailing upwind. It is the AoA that gives best L/D ratio, 10-12 degrees. When running, you use the wing as a drag device and turn it to present the widest area, more or less 90 degrees to the wind. There is a discontinuity when going from one mode to the other, and another when jibing: as the boat jibes from starboard to port tack, the sail needs to rotate 180 degrees to keep the leading edge into the wind. I'm not yet sure how to handle that--probably with some kind of hysteresis.

**Comment by rmackay9** [comment]
nice to see this additional support in sailboats.  @IamPete1 is the maintainer so I leave the decision to him but I wonder if we need sparate main_sheet and mast_rotation values.  I guess there's no way we could squash them together?

File: Rover/AP_MotorsUGV.h
Comment (arikrupnik) [reply Rover/AP_MotorsUGV.h]: You are absolutely correct. FIxed.

**Comment by arikrupnik** [comment]
@rmackay9 There was some discussion about that here: https://discuss.ardupilot.org/t/sailboat-support/32060/767

There are two sail control output types in master, this PR is proposing a third. It is unlikely that the same boat would have different types of control, even it she has multiple sails. The alternative is to have one output type, and a new parameter that would control which type of sail control the boat has. Maybe even make `SAIL_ENABLE` an enumeration, where 0 would mean disable, 1 would mean winch, 2 would be wing-with-elevator and 3 would be direct rotation control.

File: libraries/SRV_Channel/SRV_Channel.cpp
Comment (rmackay9) [inline libraries/SRV_Channel/SRV_Channel.cpp:68]: I think maybe we're getting to the point where we need to split this by vehicle so that the values can be limited to the vehicle they apply to.  This doesn't need to be a blocker for this PR though.

**Comment by rmackay9** [comment]
@arikrupnik,

Yes, I can imagine it's a choice between a servo function and a type parameter... adding a servo function is probably closer to what we are doing for other methods of control.

File: libraries/SRV_Channel/SRV_Channel.cpp
Comment (arikrupnik) [reply libraries/SRV_Channel/SRV_Channel.cpp:68]: What would you do with functions that are common to two or more types? GROUND_STEERING, for instance, makes sense on plane and rover, but not on copter.

Comment (rmackay9) [reply libraries/SRV_Channel/SRV_Channel.cpp:68]: we would need to have separate @Values{Plane} and @Values{Rover} etc.  I.e. one for each vehicle.

**Comment by arikrupnik** [comment]
I made a new parameter, `SAIL_WNDVN_ROTATES`. First, please tell me if I did this right, I probably missed some aspect of parameter definition. Second, I wasn't sure if it should go under sailboat or windvane parameters. I put it under sail, since this is the only class that can use it, seeing as no other frame has rotating masts. I can move it if it makes more sense in the other place.

The purpose of this parameter is to accommodate the use case, like my MaxiMOOP, where the vane is mounted on top of the mast and the mast rotates. In this case, the code must subtract mast rotation angle from vane reading to compute apparent wind angle. This is possible both with winch and direct-control masts, as in both cases the code attempts to set the angle of the sail.

To perform the computation, I must read the previous value of the sail servo, before the running code requests a new setting. I'm not sure how to accomplish that--perhaps @IamPete1 or @rmackay9 can point me to an example.

File: Rover/sailboat.cpp
Comment (IamPete1) [inline Rover/sailboat.cpp:110]: Your only allowed 16 chars total for a param name, these automatically get the `SAIL_` prefix added so your 2 too long

Comment (IamPete1) [reply Rover/sailboat.cpp:110]: I think it should also be in the wind-vane library, other wise logging and MAVLink will all be wrong. I don't think its going to work that well unless we add sport for precise angle measurement of the mast.

Comment (arikrupnik) [reply Rover/sailboat.cpp:110]: Maybe I leave it alone for now then and rig a mechanical solution to my use case.

Comment (IamPete1) [reply Rover/sailboat.cpp]: Feel free to give it a try, the disadvantage is that the wind reading accuracy would be the accuracy of the mast angle + the accuracy of the wing sensor. Of course if your mast angle sensor/estimate is good than that is not a problem. That is certainly something I would like to support in the future.

Comment (arikrupnik) [reply Rover/sailboat.cpp]: My assumption was that knowing the servo command is a close enough approximation for the wind angle--perhaps more so for a direct-drive mast than a winch. Of course the way you set the winch is by deriving the angle you want the boom to rotate to. But i should probably get _something_ working before I pile on features. For future reference, how can I find what was the last command we sent to a servo?

Comment (IamPete1) [reply Rover/sailboat.cpp]: `SRV_Channels::get_output_scaled(SRV_Channel::k_mast_rotation)`. But because the mounted on the mast flag should live in windvane and the mast angle stuff is a rover value it might be easier to just pass the offset to AP_WIndvane in a new function.

Comment (arikrupnik) [reply Rover/sailboat.cpp]: That was my thinking behind putting it under `SAIL`: winch and mast_rotation branches of `get_throttle_and_mainsail_out` would check the parameter and it it's true, apply the correction right there without writing the computed value anywhere.

**Comment by arikrupnik** [comment]
My vane arrived and found a temporary place on that boat that does need to know mast rotation to work. It looks funny, especially with the small sail, but it's functional. It appeals to my laziness: I simply cut a piece off the stock mount and drilled it to fit my mast step rails.

File: Rover/sailboat.cpp
Comment (IamPete1) [inline Rover/sailboat.cpp:298]: not really a fan of using nans here

File: Rover/sailboat.cpp
Comment (IamPete1) [inline Rover/sailboat.cpp:269]: i would like to see this hardcoded 90 a param, you could reuse the max sail angle.

File: Rover/sailboat.cpp
Comment (arikrupnik) [reply Rover/sailboat.cpp]: replaced with a boolean flag

File: Rover/sailboat.cpp
Comment (arikrupnik) [reply Rover/sailboat.cpp:269]: Changed to +- sail_angle_max and amended the parameter description

File: Rover/sailboat.cpp
Comment (IamPete1) [inline Rover/sailboat.cpp:279]: I think its nicer to use the tack flag from wind-vane as we do for wing elevators.

File: Rover/sailboat.cpp
Comment (IamPete1) [reply Rover/sailboat.cpp:269]: the comment it now out of date ;)

File: Rover/sailboat.cpp
Comment (IamPete1) [inline Rover/sailboat.cpp:299]:
```suggestion
mast_rotation_out = SRV_Channels::get_output_scaled(SRV_Channel::k_mast_rotation) * sail_angle_max;

```

You don't need the flag then

Comment (IamPete1) [reply Rover/sailboat.cpp:299]: You will need to set the sign/tack to match too

File: Rover/sailboat.cpp
Comment (IamPete1) [inline Rover/sailboat.cpp:289]:
```suggestion
            mast_rotation_angle = wind_dir_apparent_abs - sail_angle_ideal + pid_offset;
```
Need to add back in the ability to sheet out if were tipping over.

File: Rover/sailboat.cpp
Comment (arikrupnik) [reply Rover/sailboat.cpp:269]: Thank you for noticing this. Taken out.

File: Rover/sailboat.cpp
Comment (arikrupnik) [reply Rover/sailboat.cpp:288]: Yes. I need to account for `pid_offset`.

`mast_rotation_angle` is in degrees though, and as I recall, `pid_offset` is in percent, so I don't think you can add them here. (https://discord.com/channels/674039678562861068/797946854792757259/829132730327105576)

From reading `AR_AttitudeControl::get_sail_out_from_heel()` I _think_ the result is always zero or negative. If that is the case, you'd want to _subtract_ it from mast rotation. You want to _decrease_ the AoA, i.e. bring the sail closer to the wind to depower it. Let me know if I have this backwards.

File: Rover/sailboat.cpp
Comment (arikrupnik) [reply Rover/sailboat.cpp:298]: I'm not sure I understand how this would work. You didn't like the NaN, so I replaced it with a flag. Are you trying to get rid of the flag now? You can do that, but it would take a lot more restructuring of the logic. If I read your suggestion correctly, you want to set `mast_rotation_out`, which is in percent, in the middle of angular calculations. In addition to confusing the structure, this assignment would get overwritten in the next block, where I convert from degrees to percent. You'd need a flag to prevent that from happening, which is what I have now without mixing degrees and percent.

Your second remark is more obscure. When running, the sign is determined by hysteresis, not tack. This is the whole purpose of the special case with the `rotation_necessary` flag--to ignore the tack in this one case.

The structure of my code is like this:

* Compute optimum absolute _angle_ of attack depending on the point of sail. The calculation is the same for port and starboard tacks in all cases except dead downwind. Everything is in degrees.
* Put the sign back in. Everything is still in degrees.
* Scale from degrees to percent.
* For the special case for reducing hysteresis when running downwind, keep current output. This part is entirely in percent.

If you want me to restructure this, please let me know.

File: Rover/sailboat.cpp
Comment (arikrupnik) [reply Rover/sailboat.cpp:278]: I'd be a lot more comfortable using it if it were a function. I'm hesitant to use a stateful variable set somewhere far away, filtered in ways that are not obvious from the call site. Is `AP_WindVane::Sailboat_Tack::TACK_PORT` a statement of fact (wind is coming over the starboard side) or a statement of intent (the AP _wants_ to boat to be on a starboard tack?) Can it get out of sync with the absolute value? What guarantees do I have this will not change in a future PR?

As I've written it, with very local state, the relationship between absolute value and sign is obvious and is likely to stay consistent as the code develops.

https://discord.com/channels/674039678562861068/797946854792757259/830491529654304789

File: Rover/sailboat.cpp
Comment (IamPete1) [reply Rover/sailboat.cpp:298]: Yeah, I think you can remove the flag and simplify the code.

File: Rover/sailboat.cpp
Comment (arikrupnik) [reply Rover/sailboat.cpp:288]: I think you're right to use `pid_offset` here, only for upwind points of sail. What is this value a percent of? If it a fraction of available power that it wants me to use, like "we're heeling too much, reduce power by 30%?" If this is how it works, it may be easier to _multiply_ `sail_angle_ideal` by this fraction instead of _adding_ this value in the end. This assumes linearity between angle of attack and lift, which is shaky... or just makes this shaky assumption explicit.

Comment (IamPete1) [reply Rover/sailboat.cpp:288]: `pid_offset` is a none-dimensional output of the PID controller, you can use it however you like. You could add it to the scaled value at the end to be more consistent with the mainsail.  We add it to mainsail_base, so I think adding it here is correct too. The assumption here is of course that you only have one type of sail fitted at a time, the PIDs won't sheet out each type a equal amount.

File: Rover/sailboat.cpp
Comment (IamPete1) [reply Rover/sailboat.cpp:278]: Were asking the windvane what tack it thinks were on, its independent of navigation. Once the new wind-vane code is in this will let you filter the tack wind direction differently to the apparent wind direction.

File: Rover/sailboat.cpp
Comment (arikrupnik) [reply Rover/sailboat.cpp:288]: On a wider note, I keep tripping over the idea that AP sets sails in response to wind changes instead of adjusting the rudder. On a real boat, even a small one, sheeting is expensive in terms of both work and time, especially when there are multiple sheets. If you're about to capsize, you'd steer to fall off the wind ("tiller towards trouble"), not try to sheet out. I have to remind myself that Rover's pedigree is in powered vehicles that simply push through disturbances, but it messes me up every time I try to understand the wider logic of the code. The same happens to me when dealing with Plane. By default, Plane wants to use throttle to control airspeed, and then respond to changes in altitude with elevator, which is opposite from how you fly a full-size plane.

Comment (IamPete1) [reply Rover/sailboat.cpp:288]: Yeah, you would, that is very hard to program. In reality you would almost always do a combination of both and sheet ease and a course change. Depending on the situation and wind speed and point of sail you might do more of one or the other. Plane has the TECS system, and all its parameters to decide that, we need something similar for this. 

Its only sail at the moment because its much easier to change the sail output than to break into the steering output and clutter the code from normal rover users. We also have to stay in our box, if the user has set a heading we cant just sail at a different heading and not sheet out. Really we could only do it properly in modes where we have a target location and are free to pick the best heading to get us there.

Comment (arikrupnik) [reply Rover/sailboat.cpp:288]: It's fine as an approximation, and on little boats with full-time winch servos and relatively large batteries it works fine. (May not work as well in terms of battery capacity on Microtransat which was my original motivation for this boat). What you refer to as "a combination of both" is feed forward. If you want your airplane to climb, you increase demanded throttle directly, and use FF to tell the system it's going to need to start pulling up to maintain airspeed.

> ...break into the steering output and clutter the code from normal rover users...
> If the user has set a heading we cant just sail at a different heading...

That's the power vehicle pedigree, and I respect that. But we're not talking about choosing best course for destination and wind here. We're talking about preventing a capsize (plane equivalent is avoiding stall). In any case, this part of the conversation if outside the scope of my PR. If you ever decide to look at making rudder the primary control without messing up the rest of the codebase, I'm happy to be a sounding board.

What I need here is for you to tell me about `pid_offest`. Does it have a sign? Is it always positive? I understand that when you wrote this originally, there was a very strong assumption that a sheet works the same port and starboard, but it would be wonderful if you could document your assumptions and invariants either here or in `AR_AttitudeControl.h`. I can see that you set `p` and `i` to zero if they are positive, but not `d` or `ff`. Then you multiply the sum by -1.0f, calling it "constrain and return final output." It is probably my lack of familiarity with AP PID controllers, but by now I'm completely lost about what range of values to expect. Then in `sailboat.cpp` you multiply this value by 100 and constrain it to [0.0f..100.0f], and I have no idea what I'm looking at. Help me out?

Comment (IamPete1) [reply Rover/sailboat.cpp:288]: Yeah, you will have to swap the sign of the PID offset depending on the tack. I'm not sure off the top of my head which way round.

File: Rover/sailboat.cpp
Comment (arikrupnik) [reply Rover/sailboat.cpp:278]: I don't want to filter it differently. I want a single, consistent, atomic, signed number for apparent wind, so I can set AoA to it. Remember for a wngsail, I need to know the direction, not just magnitude of the apparent wind. Sailing with sheets is a lot simpler computationally, which is why everyone has been doing it that way since the beginning of sailing until digital computers came along.

I split the sign off for readability. I can rewrite the code so I keep apparent wind signed and replace conditions like
` if (wind_dir_apparent_abs < sail_angle_ideal)` with 
`if ((wind_dir_apparent < sail_angle_ideal) || (wind_dir_apparent > (-1.0f * sail_angle_ideal)))` and there will be no need to argue how to find the tack.

File: Rover/sailboat.cpp
Comment (arikrupnik) [reply Rover/sailboat.cpp:288]: @IamPete1, you're answering a different question from the one I'm asking.

**Comment by arikrupnik** [comment]
Took off the last commit based on feedback from @IamPete1 on discord.

File: Rover/sailboat.cpp
Comment (arikrupnik) [reply Rover/sailboat.cpp]: 19ff950c7be91ef2c5d3bd42cbf74b0160d022ad should address this feedback.

File: Rover/sailboat.cpp
Comment (arikrupnik) [reply Rover/sailboat.cpp:300]: If you want to propose a different structure for the code, I am happy to incorporate it into my PR.

File: Rover/sailboat.cpp
Comment (IamPete1) [inline Rover/sailboat.cpp:286]: Should this 90 be sail_angle_max?

Comment (arikrupnik) [reply Rover/sailboat.cpp:286]: No. This 90 is an invariant of my sail control algorithm. `sail_angle_max` is an artifact of user setup. As my comment on `@Param: ANGLE_MAX` says, you can have a setup where the linkage is capable of rotating the mast more than +-90, but the code will limit it to 90.

File: Rover/sailboat.cpp
Comment (IamPete1) [inline Rover/sailboat.cpp:289]:
```suggestion
            float power_coefficient = 1.0f - pid_offset * 0.01f;  // pid_offset==0.0f: no restrictions on power; ==100.0f: depower entirely
```
if your going to do another patch, you might as well swap out for a multiply. The comment would probably also be better up where we first get the PID offset.

File: Rover/sailboat.cpp
Comment (arikrupnik) [reply Rover/sailboat.cpp:286]: Conceivably, the use may set it up so that less than +-90 is achievable. This is suboptimal, but will still result in correct AoAs on upwind headings.

File: Rover/sailboat.cpp
Comment (arikrupnik) [reply Rover/sailboat.cpp:289]: I was going back and forth on that. In the end I decided that I was the odd man out and left this math in my piece of the code. If I were going to move this up to the initial calculation, I'd do two things:

1. Do all the constraining in `get_sail_out_from_heel()` instead of some there and some in `sailboat.cpp`
2. Make `get_sail_out_from_heel() return a value in range [0..1] instead of [0..100] and multiply it in mainsail_out and wingsail_out calculations.

I didn't do it to avoid disturbing code that has its own flow to it, and which I cannot test, having no access to boats with relevant configurations.

I'm happy to do it though if you prefer.

The division is probably no more expensive (gcc can normally figure this out when it's a constant). I used it for symmetry with multiplication by 100.0f nu line 223. Happy to change as part of the bigger change above if you want it.

Comment (arikrupnik) [reply Rover/sailboat.cpp:289]: The comment could go on `get_sail_out_from_heel()` and there would be no need for it in `sailboat.cpp`
I basically don't like upsetting other people's code unless they ask me to.

Comment (arikrupnik) [reply Rover/sailboat.cpp]: I just pushed ec8815dd24636e0f4b2c0c5fc149d76b53068064 which addresses this idea. If you think it's an improvement, I can squash it into one commit. If you think it's uglier than existing code, I'm happy to take it off.

**Review by IamPete1 — APPROVED** [review]
Looks good to me, thanks for putting up with all the change requests!

**Comment by arikrupnik** [comment]
> Looks good to me, thanks for putting up with all the change requests!

Thank you for guiding me through a first-time PR process.

**Comment by kpat0033** [comment]
Hello. Does this work with an open loop stepper motor? https://www.makerstore.com.au/product/elec-nema23-635-ht/?fbclid=IwY2xjawN9EQVleHRuA2FlbQIxMQBzcnRjBmFwcF9pZAEwAAEe22BPFGxIDq4Th-0h9ZcWfvO_3YRIyKDnDBkaY1Q98PI2mQVFAU5c3wliJU4_aem_JpGaHH-0Wzo4t4tqei2j_w
If not, what can I add and change in the software to enable it. 
Any help will be very much appreciated. We are at a real loss of how to proceed. 
Thank you!

**Comment by arikrupnik** [comment]
> Hello. Does this work with an open loop stepper motor?

There is an open issue to add support for stepper motors: https://github.com/ArduPilot/ardupilot/issues/9172

**Pull request merged**
